plenum/server/replica.py

```diff
@@ -1926,12 +1915,21 @@ def stashOutsideWatermarks(self, item: Tuple):
         self.stashingWhileOutsideWaterMarks.append(item)
 
     def processStashedMsgsForNewWaterMarks(self):
-        # `stashingWhileOutsideWaterMarks` can grow from methods called in the
-        # loop below, so `stashingWhileOutsideWaterMarks` might never
+        # Items from `stashingWhileOutsideWaterMarks` can be re-enqueued back
+        # to this queue, so `stashingWhileOutsideWaterMarks` might never
         # become empty during the execution of this method resulting
-        # in an infinite loop
+        # in an infinite loop. So we should consume each item only once
+        # during one call of this method.
         itemsToConsume = len(self.stashingWhileOutsideWaterMarks)
-        while itemsToConsume:
+
+        # Moreover, the watermarks may be updated again while consuming items
+        # from `stashingWhileOutsideWaterMarks`. So this method may be called
+        # recursively. We should stop iteration in all the outward calls of
+        # this method in case we complete consuming of stashed messages in any
+        # recursive call of this method.
+        self.consumedAllStashedMsgs = False
```

Comment:
A recursive call should be executed if it is called, because this means that the watermarks have been updated once again and all the still stashed messages should be re-considered again. The intention of this flag is to exit from the outward call(s) prematurely in case the inward call has been completed.